### ISSUE TITLE
Overload "[.ffdf" so that subsetting on logical ff vectors works

### ISSUE THREAD

**Issue by jwijffels** [edwindj/ffbase#26]
reminder
"[.ff" and "[<-.ff" work.
but also having this for ffdf would be nice.
Overload the functions from ff.

**Comment by schuemie** [comment]
Yes, this would be a nice enhancement. This works:
```r
x <- data.frame(id = c(1,2,3), value = c(3,2,1))
x[x$id == 1,]
```
but this doesn't:
```r
xFfdf <- as.ffdf(x)
xFfdf[xFfdf$id == 1, ]
```

Also, this doesn't work:
```r
which(xFfdf$id == 1)
```

Currently when I'm selecting a subset from an ffdf, I always do this:
```r
temp <- xFfdf$id == 1
xFfdf[ffbase::ffwhich(temp, temp == TRUE),]
```
Any suggestion for a more elegant solution would be appreciated

**Comment by jwijffels** [comment]
This will probably do it. Can you test it out.

require(ffbase)  
data(iris)
x <- as.ffdf(iris)

## SELECTION ("[.ffdf")
"[.ffdf" <- function(x, i, j, drop = FALSE){
  if(!missing(i) && is.ff(i) && length(i) > 0 && is.logical(i[1])){
    idx <- ffwhich(i, i == TRUE)
    if(length(idx) == 0){
      stop('ffbase is about large data, not about no data, make sure your logical indexing selects at least some data')
    }
    finalizer(idx) <- "delete"
    ff::"[.ffdf"(x, idx, j, drop = drop)
  }else{
    ff::"[.ffdf"(x, i, j, drop = drop)
  }
}
x[x$Sepal.Length > 5, ]
x[x$Sepal.Length > 5, 1:3]
x[x$Sepal.Length > 5, 1, drop=TRUE]
x[x$Sepal.Length > 5, 1]

## ASSIGNMENT ("[<-.ffdf")
testpositions <- x$Sepal.Length > 5
testpositions <- ffwhich(testpositions, testpositions == TRUE)
mynewdata <- x[testpositions, c("Sepal.Length", "Sepal.Width")]
mynewdata$Sepal.Length <- ff(1, length = nrow(mynewdata))
x[testpositions, c("Sepal.Length", "Sepal.Width")] <- mynewdata
x[testpositions, ]

"[<-.ffdf" <- function(x, i, j, value){
  if(!missing(i) && is.ff(i) && length(i) > 0 && is.logical(i[1])){
    idx <- ffwhich(i, i == TRUE)
    if(length(idx) == 0){
      stop('ffbase is about large data, not about no data, make sure your logical indexing selects at least some data')
    }
    finalizer(idx) <- "delete"
    ff::"[<-.ffdf"(x, idx, j, value)
  }else{
    ff::"[<-.ffdf"(x, i, j, value)
  }
}
data(iris)
x <- as.ffdf(iris)
x[x$Sepal.Length > 5, c("Sepal.Length", "Sepal.Width")] <- mynewdata
x[testpositions, ]

**Comment by schuemie** [comment]
Yes, that works! 

Minor suggestion: I would change the error text to something less mysterious, like "Your expression selected zero rows, but ffdf objects cannot have zero rows"

I tested it on some toy examples. If you add it to the ffbase package on Github I can test it on a larger scale. (although the code seems simple enough)

**Comment by jwijffels** [comment]
This enhancement is now available in commit 58eebe2f882d0bef5c442636a968f2b08ef2681f
Do test!

**Comment by schuemie** [comment]
I checked, and the new version seems to work and not break any of my existing code.
